File: backend/app/models/repository.py
"""Repository Intelligence Engine ORM models (Sprint 12).

Models a scanned software repository as first-class data: files, modules, symbols,
dependencies, relationships, detected architecture layers, a search index, change
history, metrics, issues, and TODOs. This is the software-understanding layer AI
employees consult before generating or reviewing code.
"""

import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.domain.repository_enums import (
    ArchitectureLayer,
    DependencyType,
    IssueSeverity,
    Language,
    ModuleKind,
    RelationshipType,
    ScanStatus,
    SymbolType,
)
from app.models.base import GUID, Base, TimestampMixin, UUIDPrimaryKeyMixin


class Repository(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "repositories"

    slug: Mapped[str] = mapped_column(String(120), nullable=False, unique=True, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    root_path: Mapped[str] = mapped_column(String(1000), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    primary_language: Mapped[str | None] = mapped_column(String(40), nullable=True)
    frameworks: Mapped[str | None] = mapped_column(Text, nullable=True)  # comma-separated
    project_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("projects.id", ondelete="SET NULL"), nullable=True
    )
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="registered")
    last_scanned_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)


class RepositoryScan(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_scans"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    status: Mapped[ScanStatus] = mapped_column(
        String(20), nullable=False, default=ScanStatus.QUEUED
    )
    file_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    symbol_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    module_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class RepositoryModule(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_modules"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    path: Mapped[str] = mapped_column(String(1000), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(300), nullable=False)
    kind: Mapped[ModuleKind] = mapped_column(String(20), nullable=False, default=ModuleKind.MODULE)
    language: Mapped[str | None] = mapped_column(String(40), nullable=True)
    file_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)


class RepositoryFile(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_files"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    module_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("repository_modules.id", ondelete="SET NULL"), nullable=True
    )
    path: Mapped[str] = mapped_column(String(1000), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(300), nullable=False)
    extension: Mapped[str | None] = mapped_column(String(40), nullable=True)
    language: Mapped[str] = mapped_column(String(40), nullable=False, default=Language.OTHER)
    size_bytes: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    line_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    symbol_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    layer: Mapped[str | None] = mapped_column(String(30), nullable=True, index=True)
    is_config: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_generated: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_test: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)


class RepositorySymbol(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_symbols"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    file_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repository_files.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(300), nullable=False, index=True)
    symbol_type: Mapped[SymbolType] = mapped_column(String(30), nullable=False, index=True)
    line: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    end_line: Mapped[int | None] = mapped_column(Integer, nullable=True)
    parent: Mapped[str | None] = mapped_column(String(300), nullable=True)
    signature: Mapped[str | None] = mapped_column(Text, nullable=True)
    docstring: Mapped[str | None] = mapped_column(Text, nullable=True)


class RepositoryDependency(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_dependencies"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    source: Mapped[str] = mapped_column(String(1000), nullable=False, index=True)
    target: Mapped[str] = mapped_column(String(1000), nullable=False, index=True)
    dependency_type: Mapped[DependencyType] = mapped_column(
        String(20), nullable=False, default=DependencyType.IMPORT
    )
    external: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)


class RepositoryRelationship(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_relationships"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    source: Mapped[str] = mapped_column(String(300), nullable=False, index=True)
    target: Mapped[str] = mapped_column(String(300), nullable=False, index=True)
    relationship_type: Mapped[RelationshipType] = mapped_column(String(20), nullable=False)


class RepositoryArchitecture(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_architecture"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    layer: Mapped[ArchitectureLayer] = mapped_column(String(30), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    file_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    symbol_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)


class RepositorySearchIndex(UUIDPrimaryKeyMixin, Base):
    """Denormalized search rows (name/kind/file). Semantic-ready seam."""

    __tablename__ = "repository_search_index"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    symbol_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("repository_symbols.id", ondelete="CASCADE"), nullable=True
    )
    file_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("repository_files.id", ondelete="CASCADE"), nullable=True
    )
    term: Mapped[str] = mapped_column(String(300), nullable=False, index=True)
    kind: Mapped[str] = mapped_column(String(30), nullable=False, index=True)
    file_path: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    line: Mapped[int | None] = mapped_column(Integer, nullable=True)


class RepositoryChangeHistory(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_change_history"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    scan_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("repository_scans.id", ondelete="SET NULL"), nullable=True
    )
    file_path: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    change_type: Mapped[str] = mapped_column(String(20), nullable=False, default="scanned")
    detail: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class RepositoryMetrics(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "repository_metrics"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    file_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    module_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    symbol_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    class_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    function_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    route_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    model_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    line_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    dependency_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    test_file_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    languages: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON string
    technical_debt: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    health_score: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)


class RepositoryIssue(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_issues"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    file_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("repository_files.id", ondelete="SET NULL"), nullable=True
    )
    severity: Mapped[IssueSeverity] = mapped_column(
        String(20), nullable=False, default=IssueSeverity.INFO
    )
    category: Mapped[str] = mapped_column(String(40), nullable=False, default="quality")
    message: Mapped[str] = mapped_column(Text, nullable=False)
    file_path: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    line: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class RepositoryTodo(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "repository_todos"

    repository_id: Mapped[uuid.UUID] = mapped_column(
        GUID(), ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True
    )
    file_id: Mapped[uuid.UUID | None] = mapped_column(
        GUID(), ForeignKey("repository_files.id", ondelete="SET NULL"), nullable=True
    )
    file_path: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    line: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    tag: Mapped[str] = mapped_column(String(20), nullable=False, default="TODO")
    text: Mapped[str] = mapped_column(Text, nullable=False)

File: frontend/src/api/repository.ts
import type { DataResponse, ListResponse } from "../types";
import { http } from "./client";

export interface RepoMetrics {
  file_count: number;
  module_count: number;
  symbol_count: number;
  class_count: number;
  function_count: number;
  route_count: number;
  model_count: number;
  line_count: number;
  dependency_count: number;
  test_file_count: number;
  technical_debt: number;
  health_score: number;
  languages: Record<string, number>;
}

export interface Repository {
  id: string;
  slug: string;
  name: string;
  root_path: string;
  description: string | null;
  primary_language: string | null;
  frameworks: string[];
  status: string;
  last_scanned_at: string | null;
  metrics: RepoMetrics | null;
}

export interface RepoFile {
  id: string;
  path: string;
  name: string;
  language: string;
  layer: string | null;
  line_count: number;
  symbol_count: number;
  is_test: boolean;
  is_config: boolean;
  is_generated: boolean;
}

export interface RepoSymbol {
  id: string;
  name: string;
  symbol_type: string;
  line: number;
  end_line: number | null;
  parent: string | null;
  signature: string | null;
  docstring: string | null;
  file_id: string;
  file_path: string | null;
}

export interface ArchLayer {
  layer: string;
  name: string;
  file_count: number;
  symbol_count: number;
  description: string | null;
}

export interface RepoModule {
  id: string;
  path: string;
  name: string;
  kind: string;
  file_count: number;
}

export interface SearchHit {
  term: string;
  kind: string;
  file_path: string | null;
  line: number | null;
  symbol_id: string | null;
  file_id: string | null;
}

export interface ImportGraph {
  nodes: { id: string; layer: string | null }[];
  edges: { source: string; target: string; external: boolean }[];
}

export interface ExternalDep {
  package: string;
  usages: number;
}

export interface Impact {
  file_path: string;
  dependencies: string[];
  dependents: string[];
  potential_breakages: string[];
  related_tests: string[];
  related_apis: { name: string; signature: string | null; line: number }[];
  related_documentation: string[];
}

export interface RepoDashboard extends Repository {
  architecture: ArchLayer[];
  external_dependencies: ExternalDep[];
  issues: { severity: string; category: string; message: string; file_path: string | null }[];
  todo_count: number;
}

export interface ScanResult {
  scan_id: string;
  status: string;
  file_count: number;
  symbol_count: number;
  module_count: number;
  duration_ms: number | null;
  summary: string | null;
  repository: Repository;
}

export const repositoryApi = {
  list: () => http.get<ListResponse<Repository>>("/repositories"),
  get: (id: string) => http.get<DataResponse<Repository>>(`/repositories/${id}`),
  register: (name: string, root_path: string, description?: string) =>
    http.post<DataResponse<Repository>>("/repositories", { name, root_path, description }),
  scan: (id: string) => http.post<DataResponse<ScanResult>>(`/repositories/${id}/scan`, {}),
  dashboard: (id: string) => http.get<DataResponse<RepoDashboard>>(`/repositories/${id}/dashboard`),
  files: (id: string, layer?: string) =>
    http.get<ListResponse<RepoFile>>(`/repositories/${id}/files${layer ? `?layer=${layer}` : ""}`),
  modules: (id: string) => http.get<ListResponse<RepoModule>>(`/repositories/${id}/modules`),
  architecture: (id: string) =>
    http.get<ListResponse<ArchLayer>>(`/repositories/${id}/architecture`),
  symbols: (id: string, symbolType?: string) =>
    http.get<ListResponse<RepoSymbol>>(
      `/repositories/${id}/symbols${symbolType ? `?symbol_type=${symbolType}` : ""}`,
    ),
  fileSymbols: (id: string, fileId: string) =>
    http.get<ListResponse<RepoSymbol>>(`/repositories/${id}/files/${fileId}/symbols`),
  references: (id: string, name: string) =>
    http.get<
      DataResponse<{
        definitions: RepoSymbol[];
        relationships: { source: string; target: string; type: string }[];
      }>
    >(`/repositories/${id}/references?name=${encodeURIComponent(name)}`),
  importGraph: (id: string) =>
    http.get<DataResponse<ImportGraph>>(`/repositories/${id}/import-graph`),
  dependencies: (id: string) =>
    http.get<ListResponse<ExternalDep>>(`/repositories/${id}/dependencies`),
  search: (id: string, q: string, kind?: string) =>
    http.get<ListResponse<SearchHit>>(
      `/repositories/${id}/search?q=${encodeURIComponent(q)}${kind ? `&kind=${kind}` : ""}`,
    ),
  impact: (id: string, filePath: string) =>
    http.get<DataResponse<Impact>>(
      `/repositories/${id}/impact?file_path=${encodeURIComponent(filePath)}`,
    ),
};

export const symbolTypeLabel = (t: string): string =>
  t.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
